### PR TITLE
KAFKA-3721: Put UpdateMetadataRequest V2 in 0.10.0-IV1

### PR DESCRIPTION
**Pull request by becketqin**

### CODE REVIEW

**Comment by gwenshap** [comment]
Do we need a documentation update?

**Comment by gwenshap** [comment]
We should also start a branchbuilder to make sure the upgrade tests didn't break

File: core/src/main/scala/kafka/api/ApiVersion.scala [hunk 1]
```diff
@@ -102,3 +103,9 @@ case object KAFKA_0_10_0_IV0 extends ApiVersion {
   val messageFormatVersion: Byte = Message.MagicValue_V1
   val id: Int = 4
 }
+
+case object KAFKA_0_10_0_IV1 extends ApiVersion {
+  val version: String = "0.10.0-IV1"
+  val messageFormatVersion: Byte = Message.MagicValue_V1
+  val id: Int = 4
```
Comment (ijuma) [inline core/src/main/scala/kafka/api/ApiVersion.scala]: This id is the same as the one before.

Comment (ijuma) [inline core/src/main/scala/kafka/api/ApiVersion.scala]: I like Gwen's suggestion about using `KAFKA_0_10_0` here as we know that will be the final version for the release, but let's see what @junrao says.

**Comment by ijuma** [comment]
From a brief look, it seems fine apart from one comment (and one question for Jun). It would be good for @junrao to take a look too. And we definitely need to run the system tests.

**Comment by junrao** [comment]
Thanks for the patch. A couple of comments.
1. We actually made another inter-broker protocol change related to SaslHandshake after the rack-awareness change. The following is how we are using it in KafkaConfig.
   
   val saslInterBrokerHandshakeRequestEnable = interBrokerProtocolVersion >= KAFKA_0_10_0_IV0

So, if we want to do this right, we will need to add 0.10.0-IV2 for SaslHandshake change and use 0.10.0-IV2 in KafkaConfig.
1. It's a bit hard to see which internal version is for which inter-broker protocol change. Could you add a comment for all the 0.10.0-IV\* to describe the corresponding protocol change?

File: core/src/main/scala/kafka/api/ApiVersion.scala (hunk 1 above)
Comment (junrao) [inline core/src/main/scala/kafka/api/ApiVersion.scala]: About using KAFKA_0_10_0 as the last version. The initial intention is to make trunk always in a releasable state by pointing KAFKA_0_10_0 to the last IV tag. I guess Gwen's suggestion is to only add KAFKA_0_10_0 just before the release and use it to replace the last IV tag? That could also work, but would mean that we need to remember doing this before every release and also make last minute code changes.

**Comment by ijuma** [comment]
@junrao since we are doing this at the wrong time, is there any benefit to adding two versions? People could be at any point in trunk at this point after all.

File: core/src/main/scala/kafka/api/ApiVersion.scala (hunk 1 above)
Comment (ijuma) [inline core/src/main/scala/kafka/api/ApiVersion.scala]: Yeah, that's the downside of that approach. I thought that we could take advantage of the fact that we're making the change right before the release, but maybe that's confusing. I vote for keeping it simple and going with the IV approach as in the initial PR.

**Comment by gwenshap** [comment]
Upgrade tests running: https://jenkins.confluent.io/job/system-test-kafka-branch-builder/452/console

**Comment by junrao** [comment]
@ijuma : Yes, you are right. It's a bit too late to fix this completely at this moment. What we can probably do is to just change the reference to KAFKA_0_10_0_IV0 in KafkaConfig to KAFKA_0_10_0_IV1. This means that KAFKA_0_10_0_IV1 will be corresponding to both the rack-awareness and the SaslHandshake change.

**Comment by ijuma** [comment]
Sounds good to me.

**Comment by becketqin** [comment]
Thanks @junrao . I updated the patch to point SASL handshake to 0.10.0-IV1 as well.
@gwenshap I just submitted a new branch builder test. Should we stop the current branch builder test?

**Comment by gwenshap** [comment]
Thanks @becketqin - I stopped mine and let yours run.

**Comment by gwenshap** [comment]
killing the system test was a REALLY bad idea.
Looks like the test nodes are now in a bad state and won't run. I'm hoping @granders is around and can help.

**Comment by gwenshap** [comment]
Retrying, just in case.... https://jenkins.confluent.io/job/system-test-kafka-branch-builder/454/console

**Comment by gwenshap** [comment]
actually... https://jenkins.confluent.io/job/system-test-kafka-branch-builder/455/console

**Comment by gwenshap** [comment]
looking better now :)

**Comment by becketqin** [comment]
Thanks, @gwenshap

**Comment by junrao** [comment]
Thanks for the patch. LGTM

**Pull request closed without merging**